src/Core/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Quantum.Simulation.Core;
+
+#nullable enable
+
+namespace Microsoft.Quantum.IQSharp
+{
+    /// <summary>
+    /// Traces through the operations in a given execution path of a Q# program by hooking on
+    /// to a simulator via the event listeners <c>OnOperationStartHandler</c> and
+    /// <c>OnOperationEndHandler</c>, and generates the corresponding <c>ExecutionPath</c>.
+    /// </summary>
+    public class ExecutionPathTracer
+    {
+        private int currDepth = 0;
+        private int renderDepth;
+        private IDictionary<int, QubitRegister> qubitRegisters = new Dictionary<int, QubitRegister>();
+        private IDictionary<int, List<ClassicalRegister>> classicalRegisters = new Dictionary<int, List<ClassicalRegister>>();
+        private List<Operation> operations = new List<Operation>();
+        private Type[] nestedTypes = new Type[]
+        {
+            typeof(Microsoft.Quantum.Canon.ApplyToEach<Qubit>),
+            typeof(Microsoft.Quantum.Canon.ApplyToEachC<Qubit>),
+            typeof(Microsoft.Quantum.Canon.ApplyToEachA<Qubit>),
+            typeof(Microsoft.Quantum.Canon.ApplyToEachCA<Qubit>),
+        };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExecutionPathTracer"/> class.
+        /// </summary>
+        /// <param name="depth">
+        /// The depth at which to render operations.
+        /// </param>
+        public ExecutionPathTracer(int depth = 1) => this.renderDepth = depth + 1;
+
+        /// <summary>
+        /// Returns the generated <c>ExecutionPath</c>.
+        /// </summary>
+        public ExecutionPath GetExecutionPath()
+        {
+            var qubits = this.qubitRegisters.Keys
+                .OrderBy(k => k)
+                .Select(k =>
+                {
+                    var qubitDecl = new QubitDeclaration(k);
+                    if (this.classicalRegisters.ContainsKey(k))
+                    {
+                        qubitDecl.NumChildren = this.classicalRegisters[k].Count;
+                    }
+
+                    return qubitDecl;
```

Comment:
Won't help directly in this case, but check out [C# 9 with-expressions](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#with-expressions).

Reply:
That's cool! I guess we could use it with something like this:
```C#
qubitDecl = qubitDecl with {
  NumChildren = (this.classicalRegisters.ContainsKey(k))
    ? this.classicalRegisters[k].Count
    : null
};
```

Reply:
Yep! Doesn't work for us yet since this project is still on C# 8, but it's exciting to keep an eye on where new features can help in the future.